### DISCUSSION TITLE
[1.4 -> 2.0]  _creation_order change

### DISCUSSION THREAD

**Discussion by Andrei-Pozolotin** [sqlalchemy/sqlalchemy#9294] [Usage Questions] [answered]
### Describe the bug

Situation: 
* the `_creation_order` in sa-1.4 and sa-2.0 is reversed to each other

Problem:
*  no support for altering column positions in `alembic` or anywhere
   https://wiki.postgresql.org/wiki/Alter_column_position

this **change of ordering** shows up in all kinds of places,  a simple example:

`_creation_order` with sa-1.4.46:
```
from sqlalchemy.dialects import postgresql as pg
stmt_default = pg.insert(ClusterSystemVerifyTBL)
assert str(stmt_default) == "INSERT INTO cluster_system_verify (ident, name, flag, data) VALUES (%(ident)s, %(name)s, %(flag)s, %(data)s)"
```

`_creation_order` with sa-2.0.3:
```
from sqlalchemy.dialects import postgresql as pg
stmt_default = pg.insert(ClusterSystemVerifyTBL)
assert str(stmt_default) == "INSERT INTO cluster_system_verify (name, flag, data, ident) VALUES (%(name)s, %(flag)s, %(data)s, %(ident)s)"
```

### SQLAlchemy Version in Use

1.4.46 and 2.0.3

### DBAPI (i.e. the database driver)

psycopg2, asyncpg,

### Database Vendor and Major Version

PostgreSQL 14

### Python Version

3.10

### Operating system

Linux

### To Reproduce

```python
see above
```

### Error

see above

**Comment by CaselIT** [comment]
Hi

There is nothing actionable in your issue report, so I converted to discussion.
At the very least you could have included what `ClusterSystemVerifyTBL` is and how it was defined

**Comment by Andrei-Pozolotin** [comment]
@CaselIT Federico: thank you for your response.

Additional information, as requested:

<br/>

**Sample Table** definition:

<details>
<summary>ClusterSystemVerifyTBL</summary>

```python

import sqlalchemy as sa
from sqlalchemy.dialects import postgresql as pg
from sqlalchemy.orm import declarative_base


class ModelDBMS:
    ""

    metadata = sa.MetaData(
    )

    tablebase = declarative_base(
        metadata=metadata,
    )


class TableIDENT:

    # note:
    # pglogical: intercept/update on target node
    # alembic: ignore override values by alembic autogenerate
    # https://github.com/sqlalchemy/alembic/issues/821
    ident = sa.Column(

        pg.BIGINT,
        server_default=sa.Identity(
            always=True,
            on_null=None,
            start=None,
            increment=1,
            # must override in server config
            minvalue=2,  # not 1, to avoid defaults in backup
            # must override in server config
            maxvalue=999999999999998,  
            nominvalue=None,
            nomaxvalue=None,
            cycle=True,
            cache=1,
            order=None,
        ),

        primary_key=True,
        nullable=False,
    )


class ClusterSystemVerifyTBL(ModelDBMS.tablebase, TableIDENT):
    "replicated table for tesing"

    __tablename__ = 'cluster_system_verify'

    name = sa.Column(sa.TEXT, unique=True)
    flag = sa.Column(sa.BOOLEAN)
    data = sa.Column(sa.INTEGER)


```

</details>

<br/>

**The problem at hand** is that in the upgrade scenario:
* using `sa` +  `alembic`
* not changing any `py` code nor `alembic` config
* and changing only `sa` version `1.4 => 2.0`

results in column creation order change in `CreateTable` stanza produced by `sa` +  `alembic` :
* specifically, when using `declarative_base` and table inheritance
* as shown in the example above of `ClusterSystemVerifyTBL`:
* the column of interest: `ident` used to be **first**, now this column is **last** in pg file table store

<details>
<summary>ClusterSystemVerifyTBL in sa-2.0 as shown by pg-admin</summary>

![image](https://user-images.githubusercontent.com/1622151/218464747-e8c129e4-06af-4818-927b-685ece373483.png)

</details>

as far as I can infer from reading `sa` code:
* usage of `_creation_order` is the culprit

<br/>

**This becomes rather problematic** 
* for wide tables in multi-master replicated use case:
https://www.cybertec-postgresql.com/en/column-order-in-postgresql-does-matter/
* and is a breaking change with existing code base / table schema

**Comment by zzzeek** [comment]
the change in behavior occurred in e545298e35ea9f126054b337e but there was also a later rework to column ordering as part of #8705 .

the old behavior can be restored (at least to the degree the above mapping indicates) using this patch:

```diff
diff --git a/lib/sqlalchemy/orm/decl_base.py b/lib/sqlalchemy/orm/decl_base.py
index aeed9b4395..1271b59c8c 100644
--- a/lib/sqlalchemy/orm/decl_base.py
+++ b/lib/sqlalchemy/orm/decl_base.py
@@ -1299,6 +1299,7 @@ class _ClassScanMapperConfig(_MapperConfig):
                                 )
 
                     column_copies[obj] = copy_ = obj._copy()
+                    copy_._creation_order = obj._creation_order
 
                     locally_collected_attributes[name] = copy_
                     setattr(cls, name, copy_)
@@ -1559,6 +1560,10 @@ class _ClassScanMapperConfig(_MapperConfig):
 
         manager = attributes.manager_of_class(cls)
 
+        declared_columns = self.declared_columns = sorted(
+            declared_columns, key=lambda c: c._creation_order
+        )
+
         if "__table__" not in clsdict_view and table is None:
             if hasattr(cls, "__table_cls__"):
                 table_cls = cast(

```

which then breaks tests that were changed in e545298e35ea9f126054b .  it's not clear yet the degree to which e545298e35ea9f126054b vs b2473649e06c1d6d29baeffb1f6ca3d75a1 which was for #8705 rely on the newer version of this change.

The general intent was that the column ordering would honor the main class first, so that a mapping like this:

```py
        class Foo:

            col1 = Column(Integer)
            col3 = Column(Integer)

        class Bar:

            col2 = Column(Integer)
            col4 = Column(Integer)

        class Model(Base, Foo, Bar):

            id = Column(Integer, primary_key=True)
            __tablename__ = "model"

```

places "id" first.

you can see there's no one way people might want to do this, in your case, you have PKs on your mixin.  

changing this hardcoded behavior back and forth is not going to be a solution.    A system by which the ordering can be controlled imperatively should be used.  We do have a hook that you can use for this right now, which is https://docs.sqlalchemy.org/en/20/orm/declarative_config.html#table-cls .  Here's usage that you can use right now to migrate to 2.0 (will work in 1.4 as well)

```py
from sqlalchemy import Table

class TableIDENT:

    # note:
    # pglogical: intercept/update on target node
    # alembic: ignore override values by alembic autogenerate
    # https://github.com/sqlalchemy/alembic/issues/821
    ident = sa.Column(

        pg.BIGINT,
        server_default=sa.Identity(
            always=True,
            on_null=None,
            start=None,
            increment=1,
            # must override in server config
            minvalue=2,  # not 1, to avoid defaults in backup
            # must override in server config
            maxvalue=999999999999998,
            nominvalue=None,
            nomaxvalue=None,
            cycle=True,
            cache=1,
            order=None,
        ),
        primary_key=True,
        nullable=False,
        info={"order_first": True}
    )


    @classmethod
    def __table_cls__(cls, name, metadata_obj, *arg, **kw):
        arg = sorted(arg, key=lambda obj: obj.info.get("order_first", 10))

        return Table(name, metadata_obj, *arg, **kw)

class ClusterSystemVerifyTBL(TableIDENT, ModelDBMS.tablebase):
    "replicated table for tesing"

    __tablename__ = 'cluster_system_verify'

    name = sa.Column(sa.TEXT, unique=True)
    flag = sa.Column(sa.BOOLEAN)
    data = sa.Column(sa.INTEGER)

from sqlalchemy.schema import CreateTable
print(CreateTable(ClusterSystemVerifyTBL.__table__))
```

the program on 1.4 and main/2.0 is ordered the same way:

```
CREATE TABLE cluster_system_verify (
	ident BIGINT GENERATED ALWAYS AS IDENTITY (INCREMENT BY 1 MINVALUE 2 MAXVALUE 999999999999998 CACHE 1 CYCLE), 
	name TEXT, 
	flag BOOLEAN, 
	data INTEGER, 
	PRIMARY KEY (ident), 
	UNIQUE (name)
)


```

**Reply by zzzeek** [reply]
im adding notes to the whats new doc right now as I'm not certain this change was documented prominently enough  (nobody referred towards it during many months of beta releases).

**Reply by zzzeek** [reply]
this is documented in 037884edaef8638c7128ad13c .  I will propose column ordering for ``mapped_column`` ( as this is ORM / Declarative specific)

**Comment by Andrei-Pozolotin** [comment]
@zzzeek Mike: thank you very much for your complete and clear answer.

We applied your `__table_cls__` pattern as above, and that resolves the issue.

Few notes, as a reflection of this little experience:

<br/>

**Yes, `__table_cls__()` needs** 
much more profound place in the documentation

<br/>

**It would help first-time users** 
of `__table_cls__()` to have type hints, i.e.:
```python
    @classmethod
    def __table_cls__(cls,
            table_name:str,
            table_metadata:sa.MetaData,
            *column_list:list[sa.Column],
            **table_kwargs:dict[str, any],  # dialect specific, and passed in the form <dialectname>_<argname>
        ) -> sa.Table:
        pass
```

<br/>

**Using `__table_cls__()` for column**  order definition
still fells like a hack, we are looking forward to the explicit `mapped_column` approach.

**Comment by Andrei-Pozolotin** [comment, marked as answer]
@CaselIT @zzzeek : Mike, Federico: 

thank your guys for:
#9297 
https://github.com/sqlalchemy/sqlalchemy/commit/361beb0bc8392c92403ffc1999eb2a9847e945c7
https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#change-9297

`mapped_column.sort_order` works great and feels natural
